File: algorithms/java/96_unique_binary_search_trees/unique_binary_search_trees.java
import java.util.Scanner;

import tree.BinTree;

/**
 * @author Peng
 *
 * Problem Description:
 *      Given n, how many structurally unique BST's (binary search trees) that store values 1...n?
 *    For example,
 *       Given n = 3, there are a total of 5 unique BST's. 
 *       
 *          1         3     3      2      1
 *           \       /     /      / \      \
 *            3     2     1      1   3      2
 *           /     /       \                 \
 *          2     1         2                 3 
 */
public class unique_binary_search_trees {

    /**
     * Solution 1:
     *  derivation as:
     *      set N(n) as the target tree number.
     *          here we count i as the root value of the tree.
     *          so N(i,n) = N(i-1) * N(n-i).
     *              while N(i-1) refers to left subtree count;
     *                    N(n-i) refers to right subtree count;
     *          N(n) = N(1,n) + N(2,n) +...+ N(n,n);
     *      so:
     *      ini:N(0) = 1, N(1) = 1;
     *     then:N(2) = N(1,2) + N(2,2) = (N(0) * N(1)) + (N(1) * N(0)) 
     *               = f(N(0), N(1))
     *          N(3) = N(1,3) + N(2,3) + N(3,3) = (N(0) * N(2)) + (N(1) * N(1)) + (N(2) * N(0))
     *               = f(N(0), N(1), N(2))
     *          ...
     * @param n
     * @return
     */
    public static int numTrees(int n) {
        int [] N = new int[n+1]; //0->n
        N[0] = 1;
        N[1] = 1;
        for(int i = 2; i <= n; i++){
            for(int j = 1; j <= i; j++){
                N[i] += N[j-1] * N[i-j];
            }
        }
        return N[n];
    }
    
    public static void main(String[] args) {
        // creating a binary tree.
        Integer[] data = {1, 2, 3, 4, 5, null, null};
        BinTree bt = new BinTree();
        bt.bTreeRoot = bt.createTree(data, 0, 7);  
        
        Scanner scanner = new Scanner(System.in);
        while(scanner.hasNext()){
            int n = scanner.nextInt();
            System.out.println("the tree number of " + n + " is: " + numTrees(n));
        }
        
        scanner.close();
    }

}

File: algorithms/python/96_unique_binary_search_trees/unique_binary_search_trees.py
"""
@author: Peng

@problem_info:    
    Given n, how many structurally unique BST's (binary search trees) that store values 1...n?

    For example,
        Given n = 3, there are a total of 5 unique BST's.
        
           1         3     3      2      1
            \       /     /      / \      \
             3     2     1      1   3      2
            /     /       \                 \
           2     1         2                 3

"""

class Solution(object):
    
    # solution 1
    def numTrees_1(self, n):
        """
        :idea: the result is "Catalan Number"
        
            set all number is [a1, a2, a3, ..., ak, ..., an], 
            the possible BST number is f(n)
            if the root is ak, then:
                the left sub_BST is [a1, ..., ak-1], 
                the right sub_BST is [ak+1, ..., an], 
            two sub_trees are mutually independent. 
            set their possible BST number set to f(k-1) and f(n-k)
            so :
                f(n) = sum_k { f(k-1)*f(n-k) }
            which is "Catalan Number"
            
        :param n: int, the number of BST nodes
        :return: int, the number of unique BSTs
        """
        fact_n = 1  # n! 
        fact_2n = 1  # (2n)!/(n+1)!
        for i in range(2, n+1): fact_n *= i
        for i in range(n+2,2*n+1): fact_2n *= i
        return int(fact_2n/fact_n)

    # solution 2
    def numTrees_2(self, n):
        """
        :idea: using DP
            the recursion equation is :
                f(n) = sum_k { f(k-1)*f(n-k) }
            here we use a list to store f(i) as i from 1 to n. (set f[0] = 1) 
            
        :param n: int, the number of BST nodes
        :return: int, the number of unique BSTs
        """
        f = []
        f.append(1)
        for i in range(1, n+1) :
            f_i = 0
            for k in range(1, i+1) :
                f_i += f[k-1] * f[i-k]
            f.append(f_i)
        return f[n]

# test code
if __name__ == '__main__':
    s = Solution()
    n = 5
    print("The number of all possible unique BST is: ", s.numTrees_2(n))
    
# - PY131 - #